shared/tools/base.py

```
@@ -0,0 +1,385 @@
+            LLMの応答
+        """
+        for attempt in range(self.max_retries):
+            try:
+                response = await asyncio.to_thread(self.llm.invoke, prompt)
+                return response.content.strip()
+            except Exception as e:
+                if attempt == self.max_retries - 1:
+                    raise AgriAIException(
+                        f"LLM呼び出しが{self.max_retries}回失敗しました: {str(e)}",
+                        category=ErrorCategory.EXTERNAL_API
+                    )
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)